Add toolrank optional support

louisroyer opened the pull request.

louisroyer committed:
Add toolrank support

louisroyer committed:
Make bow shot able to level up the bow

paly2 commented on init.lua:
```diff
@@ -363,4 +366,10 @@ function throwing.register_bow(name, def)
 	end
 	minetest.register_tool(name, def)
+	if use_toolranks then
+		minetest.override_item(name:sub(2), {
```
This assumes that the item name as passed to the function always starts with a `:` prefix, which is not necessarily true. A check should be made to remove the prefix if it exists, but keep the whole name otherwise.

Edit: wait, what is the point of overriding the item immediately after registering it? The definition of the item should be changed before registration, so as to make overriding unnecessary.

louisroyer replied:
Oh, you are right, fixed it.

paly2 commented on init.lua:
```diff
@@ -345,6 +345,9 @@ function throwing.register_bow(name, def)
 				if not minetest.settings:get_bool("creative_mode") then
 					itemstack:add_wear(65535 / (def.uses or 50))
+					if use_toolranks then
+						toolranks.new_afteruse(itemstack, user, nil, {wear= 65535 / (def.uses or 50)})
```
As this expression is used twice, it may make sense to use an auxiliary variable:
```lua
local uses = def.uses or 50
```
so that the default value for `uses` is hard-coded only once.

paly2 commented on init.lua:
```diff
@@ -7,7 +7,7 @@ throwing.target_node = 2
 
 throwing.modname = minetest.get_current_modname()
-
+local use_toolranks = minetest.get_modpath("toolranks")
```
An additional line break would be appreciable here ;-)

paly2 commented:
Good idea, thanks! Except for the minor adjustments needed before merging described in my review, it looks good to me

paly2 commented on init.lua:
```diff
@@ -363,4 +366,10 @@ function throwing.register_bow(name, def)
 		return itemstack
 	end
 	minetest.register_tool(name, def)
+	if use_toolranks then
```
It is probably undesirable to enforce the use of `toolranks` for all items. An optional bow definition boolean parameter `no_toolranks` defaulting to `false` would be better, so that the use of toolranks can be disabled when registering a bow!

louisroyer committed:
Add requested changes

louisroyer committed:
Change enable_toolranks to no_toolranks

paly2 commented on README.md:
````diff
@@ -26,7 +26,7 @@ throwing.bow_cooldown = 0.2
 There are two available functions in the mod API:
 ```lua
-function throwing.register_bow(name, definition)
+function throwing.register_bow(name, definition, no_toolranks)
````
Sorry, I meant to add the `no_toolranks` parameter as a value in the `definition` table rather than as a new function parameter: this makes using the function and reading a definition easier. All changes are otherwise perfect!

louisroyer replied:
b79ea852c947be13f6d5e12fb65f32aee606b3ea
Is it better now? Also put it in a variable to avoid duplication.

louisroyer committed:
Remove the overwritting

paly2 commented on init.lua:
```diff
@@ -344,7 +345,11 @@ function throwing.register_bow(name, def)
+					local uses = 65535 / (def.uses or 50)
+					itemstack:add_wear(uses)
+					if use_toolranks and not no_toolranks then
+						toolranks.new_afteruse(itemstack, user, nil, {wear= uses})
```
A minor code style improvement could be made here by adding a space before the `=` sign:
```lua
{wear = uses}
```

louisroyer commented:
This probably needs to be squashed before a merge.

louisroyer committed:
no_toolranks in defs in throwing.register_bow

paly2 committed:
Toolranks support: minor changes
Reword documentation in README.md, fix a typo in init.lua and enable toolranks even if def.original_description exists---but do not overwrite it.

paly2 commented:
Thanks, however, another problem arises. When the arrow (or the bow, if the bow throws itself) encounters a problem during its flight, e.g. if the arrows hits a node during its first step (as happens, for example, when a player throws an arrow inside a node), the arrow is removed and put back in the player's inventory.
This means that the number of tool uses as counted by toolranks is incremented even if the flight of the arrow was aborted (and the arrow, having been put back in the player inventory, can be thrown again), allowing a player to easily increase the number of uses of a bow by repeatedly throwing the same arrow in situation where the launch fails.

In order to address this, the tool uses must be incremented when the arrow hits its target rather than when it is thrown. If you wish, I can do this myself.

louisroyer commented:
> In order to address this, the tool uses must be incremented when the arrow hits its target rather than when it is thrown. If you wish, I can do this myself.
Thank you, I didn't know about this behavior.

paly2 committed:
Wait until the arrow hits its target before updating bow toolranks uses

paly2 commented:
I think it works, I still have to do a little more testing before merging

SmallJoker commented:
Needs rebase.

paly2 committed:
Add a throwing.toolranks settting to allow globally disabling toolranks

paly2 commented:
Rebased !
I added a global setting `throwing.toolranks` defaulting to true to allow disabling toolranks.
I believe it's ready for merging
